onnxruntime/python/tools/quantization/fp16_converter.py

```diff
@@ -0,0 +1,359 @@
+                # add Cast node (from tensor(float16) to tensor(float) before graph output
+                new_value_info = model.graph.value_info.add()
+                new_value_info.CopyFrom(n)
+                new_value_info.name = input_name
+                new_value_info.type.tensor_type.elem_type = onnx_proto.TensorProto.FLOAT16
+                new_node = [helper.make_node("Cast", [input_name], [n.name], to=1, name=node_name)]
+                model.graph.node.extend(new_node)
+                value_info_list.append(new_value_info)
+                io_casts.add(node_name)
+
+    while queue:
+        next_level = []
+        for q in queue:
+            # if q is model, push q.graph (GraphProto)
+            if isinstance(q, onnx_proto.ModelProto):
+                next_level.append(q.graph)
+            # if q is model.graph, push q.node.attribute (AttributeProto)
+            if isinstance(q, onnx_proto.GraphProto):
+                for n in q.node:
+                    # if n is in the block list (doesn't support float16), no conversion for the node,
+                    # and save the node for further processing
+                    if n.name in io_casts:
+                        continue
+                    for i in range(len(n.input)):
+                        if n.input[i] in name_mapping:
+                            n.input[i] = name_mapping[n.input[i]]
+                    for i in range(len(n.output)):
+                        if n.output[i] in name_mapping:
+                            n.output[i] = name_mapping[n.output[i]]
+                    # don't add the attr into next_level for the node in node_keep_data_type_list
+                    # so it will not be converted to float16
+                    if n.op_type not in op_allow_list:
+                        node_list.append(n)
+                    else:
+                        if n.op_type == "Cast":
+                            for attr in n.attribute:
+                                if attr.name == "to" and attr.i == 1:
+                                    attr.i = 10
+                                    break
+                        for attr in n.attribute:
+                            next_level.append(attr)
+            # if q is model.graph.node.attribute, push q.g and q.graphs (GraphProto)
+            # and process node.attribute.t and node.attribute.tensors (TensorProto)
+            if isinstance(q, onnx_proto.AttributeProto):
+                next_level.append(q.g)
+                for n in q.graphs:
+                    next_level.append(n)
+                q.t.CopyFrom(convert_tensor_float_to_float16(q.t, min_positive_val, max_finite_val))
+                for n in q.tensors:
+                    n = convert_tensor_float_to_float16(n, min_positive_val, max_finite_val)
+            # if q is graph, process graph.initializer(TensorProto), input, output and value_info (ValueInfoProto)
+            if isinstance(q, onnx_proto.GraphProto):
```

Comment:
This looks the same with line 192? What gives?

Reply:
This one is for initializer

Reply:
how can the program tell the difference? it's all a top level branch from the loop that takes out the queue items.